src/encode.rs

```diff
@@ -133,60 +137,72 @@ pub fn encode_program_no_witness<W: io::Write, App: Application>(
 /// Encode witness data to bits.
 ///
 /// Returns the number of written bits.
-pub fn encode_witness<W: io::Write>(witness: &[Value], w: &mut BitWriter<W>) -> io::Result<usize> {
+pub fn encode_witness<'a, W: io::Write, I>(witness: I, w: &mut BitWriter<W>) -> io::Result<usize>
+where
+    I: Iterator<Item = &'a Value> + Clone,
+{
     let start_n = w.n_total_written();
+    let mut witness = witness.peekable();
 
-    if witness.is_empty() {
+    if witness.peek().is_none() {
         w.write_bit(false)?;
     } else {
         w.write_bit(true)?;
-
-        let mut payload = Vec::new();
-        let mut payload_w = BitWriter::from(&mut payload);
         let mut bit_len = 0;
+        let witness_clone = witness.clone();
 
         for value in witness {
```

Comment:
In 74ff5ee08b4de0fa948b2bf4b73522fb0162cc28:

I think it'd be clearer to just do `for value in witness.clone() {' here rather than having a separate cloned variable that you keep in scope the whole time for use later.